File: src/GALERKIN/processing/ScratchVisibilityStrategy.cpp
/**
Scratch renderer routines. Used for handling intra-cluster visibility
with a Z-buffer visibility algorithm in software
*/

#include "GALERKIN/processing/ClusterTraversalStrategy.h"
#include "GALERKIN/processing/ScratchVisibilityStrategy.h"
#include "GALERKIN/processing/visitors/ScratchRendererVisitor.h"
#include "scene/Camera.h"

/**
Src is a toplevel surface element. Render the corresponding patch
with pixel value a pointer to the element. Uses global variable
eyePoint for backface culling
*/
static Vector3D globalEyePoint;

/**
Create a scratch software renderer for various operations on clusters
*/
void
ScratchVisibilityStrategy::scratchInit(GalerkinState *galerkinState) {
    galerkinState->scratch = new SGL_CONTEXT(galerkinState->scratchFrameBufferSize, galerkinState->scratchFrameBufferSize);
    GLOBAL_sgl_currentContext->sglDepthTesting(true);
}

/**
Terminates scratch rendering
*/
void
ScratchVisibilityStrategy::scratchTerminate(GalerkinState *galerkinState) {
    if ( galerkinState->scratch != nullptr ) {
        delete galerkinState->scratch;
        galerkinState->scratch = nullptr;
    }
}

/**
Sets up an orthographic projection of the cluster as
seen from the eye. Renders the element pointers to the elements
in cluster in the scratch frame buffer and returns pointer to a bounding box
containing the size of the virtual screen. The cluster nicely fits
into the virtual screen
*/
BoundingBox *
ScratchVisibilityStrategy::scratchRenderElements(GalerkinElement *cluster, Vector3D eye, GalerkinState *galerkinState) {
    static BoundingBox boundingBox;

    if ( cluster->id == galerkinState->lastClusterId &&
         eye.equals(galerkinState->lastEye, Numeric::EPSILON_FLOAT) ) {
        return &boundingBox;
    }

    // Cache previously rendered cluster and eye point in order to
    // avoid re-rendering the same situation next time
    galerkinState->lastClusterId = cluster->id;
    galerkinState->lastEye = eye;

    const Vector3D center = cluster->midPoint();
    Vector3D up = {0.0, 0.0, 1.0};
    Vector3D viewDirection;

    viewDirection.subtraction(center, eye);
    viewDirection.normalize(Numeric::EPSILON_FLOAT);
    if ( java::Math::abs(up.dotProduct(viewDirection)) > 1.0 - Numeric::EPSILON ) {
        up.set(0.0, 1.0, 0.0);
    }

    const Matrix4x4 lookAt = Matrix4x4::createLookAtMatrix(eye, center, up);

    Camera::transformBoundingBox(cluster->geometry->getBoundingBox(), lookAt, &boundingBox);

    SGL_CONTEXT *prev_sgl_context = sglMakeCurrent(galerkinState->scratch);

    Matrix4x4 o = Camera::projectionMatrixFromBoundingBox(boundingBox);
    GLOBAL_sgl_currentContext->sglLoadMatrix(&o);
    GLOBAL_sgl_currentContext->sglMultiplyMatrix(&lookAt);

    // Choose a viewport depending on the relative size of the smallest
    // surface element in the cluster to be rendered
    int vp_size = static_cast<int>((boundingBox.dx() * boundingBox.dy()) / cluster->minimumArea);

    if ( vp_size > galerkinState->scratch->width ) {
        vp_size = galerkinState->scratch->width;
    }
    if ( vp_size < 32 ) {
        vp_size = 32;
    }
    GLOBAL_sgl_currentContext->sglViewport(0, 0, vp_size, vp_size);

    // Render element pointers in the scratch frame buffer
    globalEyePoint = eye; // Needed for backface culling test
    GLOBAL_sgl_currentContext->sglClear(static_cast<SGL_PIXEL>(0x00), SGL_MAXIMUM_Z);

    ScratchRendererVisitor leafVisitor(globalEyePoint);
    ClusterTraversalStrategy::traverseAllLeafElements(&leafVisitor, cluster, galerkinState);

    sglMakeCurrent(prev_sgl_context);

    return &boundingBox;
}

/**
After rendering element pointers in the scratch frame buffer, this routine
computes the average radiance of the virtual screen
*/
ColorRgb
ScratchVisibilityStrategy::scratchRadiance(const GalerkinState *galerkinState) {
    ColorRgb rad;

    rad.clear();
    int nonBackGround = 0;
    for ( int j = 0; j < galerkinState->scratch->vp_height; j++ ) {
        const int rowStart = j * galerkinState->scratch->width;
        for ( int i = 0; i < galerkinState->scratch->vp_width; i++ ) {
            const GalerkinElement *element = reinterpret_cast<GalerkinElement *>(galerkinState->scratch->frameBuffer[rowStart + i]);
            if ( element != nullptr ) {
                if ( galerkinState->galerkinIterationMethod == GalerkinIterationMethod::GAUSS_SEIDEL ||
                     galerkinState->galerkinIterationMethod == GalerkinIterationMethod::JACOBI ) {
                    rad.add(rad, element->radiance[0]);
                } else {
                    rad.add(rad, element->unShotRadiance[0]);
                }
                nonBackGround++;
            }
        }
    }
    if ( nonBackGround > 0 ) {
        rad.scale(1.0f / static_cast<float>(galerkinState->scratch->vp_width * galerkinState->scratch->vp_height));
    }
    return rad;
}

/**
Computes the number of non background pixels
*/
int
ScratchVisibilityStrategy::scratchNonBackgroundPixels(const GalerkinState *galerkinState) {
    int nonBackGround = 0;

    for ( int j = 0; j < galerkinState->scratch->vp_height; j++ ) {
        const int rowStart = j * galerkinState->scratch->width;
        for ( int i = 0; i < galerkinState->scratch->vp_width; i++ ) {
            const GalerkinElement *elem = reinterpret_cast<GalerkinElement *>(galerkinState->scratch->frameBuffer[rowStart + i]);
            if ( elem ) {
                nonBackGround++;
            }
        }
    }
    return nonBackGround;
}

/**
Counts the number of pixels occupied by each element. The result is
accumulated in the tmp field of the elements. This field should be
initialized to zero before
*/
void
ScratchVisibilityStrategy::scratchPixelsPerElement(const GalerkinState *galerkinState) {
    for ( int i = 0; i < galerkinState->scratch->vp_height; i++ ) {
        const int rowStart = i * galerkinState->scratch->width;
        for ( int j = 0; j < galerkinState->scratch->vp_width; j++ ) {
            GalerkinElement *elem = reinterpret_cast<GalerkinElement *>(galerkinState->scratch->frameBuffer[rowStart + j]);
            if ( elem != nullptr ) {
                elem->scratchVisibilityUsageCounter++;
            }
        }
    }
}

File: typescript/src/main/vsdk/toolkit/galerkin/processing/ScratchVisibilityStrategy.ts
import { ColorRgb } from "../../common/color/ColorRgb";
import { Matrix4x4 } from "../../common/linealAlgebra/Matrix4x4";
import { Numeric } from "../../common/linealAlgebra/Numeric";
import { Vector3D } from "../../common/linealAlgebra/Vector3D";
import { GalerkinElement } from "../GalerkinElement";
import { GalerkinIterationMethod } from "../GalerkinIterationMethod";
import { GalerkinState } from "../GalerkinState";
import { ScratchRendererVisitor } from "./visitors/ScratchRendererVisitor";
import { SglConstants } from "../../render/sgl/SglConstants";
import { SglContext } from "../../render/sgl/SglContext";
import { Camera } from "../../scene/Camera";
import { BoundingBox } from "../../skin/AxisAlignedBoundingBox";
import { Element } from "../../environment/geometry/elements/Element";
import { ClusterTraversalStrategy } from "./ClusterTraversalStrategy";

export class ScratchVisibilityStrategy {
  public static scratchInit(galerkinState: GalerkinState): void {
    galerkinState.scratch = new SglContext(galerkinState.scratchFrameBufferSize, galerkinState.scratchFrameBufferSize);
    galerkinState.scratch.sglDepthTesting(true);
  }

  public static scratchTerminate(galerkinState: GalerkinState): void {
    if (galerkinState.scratch !== null) {
      galerkinState.scratch = null;
    }
  }

  public static scratchRenderElements(cluster: GalerkinElement, eye: Vector3D, galerkinState: GalerkinState): BoundingBox {
    const boundingBox = new BoundingBox();

    if (cluster.id === galerkinState.lastClusterId
      && eye.equals(galerkinState.lastEye, Numeric.EPSILON_FLOAT)) {
      return boundingBox;
    }

    galerkinState.lastClusterId = cluster.id;
    galerkinState.lastEye = eye;

    const center = cluster.midPoint();
    const up = new Vector3D(0.0, 0.0, 1.0);
    const viewDirection = new Vector3D();

    viewDirection.subtraction(center, eye);
    viewDirection.normalize(Numeric.EPSILON_FLOAT);
    if (globalThis.Math.abs(up.dotProduct(viewDirection)) > 1.0 - Numeric.EPSILON) {
      up.set(0.0, 1.0, 0.0);
    }

    const lookAt = Matrix4x4.createLookAtMatrix(eye, center, up);

    Camera.transformBoundingBox((cluster.geometry as NonNullable<GalerkinElement["geometry"]>).getBoundingBox(), lookAt, boundingBox);

    const scratch = galerkinState.scratch as SglContext;
    const o = Camera.projectionMatrixFromBoundingBox(boundingBox);
    scratch.sglLoadMatrix(o);
    scratch.sglMultiplyMatrix(lookAt);

    let vpSize = globalThis.Math.trunc((boundingBox.dx() * boundingBox.dy()) / cluster.minimumArea);
    if (vpSize > scratch.width) {
      vpSize = scratch.width;
    }
    if (vpSize < 32) {
      vpSize = 32;
    }
    scratch.sglViewport(0, 0, vpSize, vpSize);

    scratch.sglClear(0x00, SglConstants.SGL_MAXIMUM_Z);

    const leafVisitor = new ScratchRendererVisitor(eye, scratch);
    ClusterTraversalStrategy.traverseAllLeafElements(leafVisitor, cluster, galerkinState);

    return boundingBox;
  }

  public static scratchRadiance(galerkinState: GalerkinState): ColorRgb {
    const rad = new ColorRgb();
    rad.clear();
    let nonBackGround = 0;

    for (let j = 0; j < (galerkinState.scratch as SglContext).vp_height; j++) {
      const rowStart = j * (galerkinState.scratch as SglContext).width;
      for (let i = 0; i < (galerkinState.scratch as SglContext).vp_width; i++) {
        const elementBase = (galerkinState.scratch as SglContext).galerkinElementBuffer[rowStart + i];
        if (elementBase instanceof GalerkinElement) {
          const element = elementBase as GalerkinElement;
          if (galerkinState.galerkinIterationMethod === GalerkinIterationMethod.GAUSS_SEIDEL
            || galerkinState.galerkinIterationMethod === GalerkinIterationMethod.JACOBI) {
            rad.add(rad, (element.radiance as ColorRgb[])[0]);
          }
          else {
            rad.add(rad, (element.unShotRadiance as ColorRgb[])[0]);
          }
          nonBackGround++;
        }
      }
    }
    if (nonBackGround > 0) {
      rad.scale(1.0 / ((galerkinState.scratch as SglContext).vp_width * (galerkinState.scratch as SglContext).vp_height));
    }
    return rad;
  }

  public static scratchNonBackgroundPixels(galerkinState: GalerkinState): number {
    let nonBackGround = 0;

    for (let j = 0; j < (galerkinState.scratch as SglContext).vp_height; j++) {
      const rowStart = j * (galerkinState.scratch as SglContext).width;
      for (let i = 0; i < (galerkinState.scratch as SglContext).vp_width; i++) {
        const elementBase = (galerkinState.scratch as SglContext).galerkinElementBuffer[rowStart + i];
        if (elementBase !== null) {
          nonBackGround++;
        }
      }
    }
    return nonBackGround;
  }

  public static scratchPixelsPerElement(galerkinState: GalerkinState): void {
    for (let i = 0; i < (galerkinState.scratch as SglContext).vp_height; i++) {
      const rowStart = i * (galerkinState.scratch as SglContext).width;
      for (let j = 0; j < (galerkinState.scratch as SglContext).vp_width; j++) {
        const elementBase = (galerkinState.scratch as SglContext).galerkinElementBuffer[rowStart + j];
        if (elementBase instanceof GalerkinElement) {
          const elem = elementBase as GalerkinElement;
          elem.scratchVisibilityUsageCounter++;
        }
      }
    }
  }
}
